src/aliceVision/image/curves.cpp

```diff
@@ -0,0 +1,250 @@
+#include <cmath>
+
+#include "curves.hpp"
+
+
+namespace alicevision {
+    namespace image {
+
+        // Tone curve according to Adobe's reference implementation
+        // values in 0xffff space
+        void AdobeToneCurve::Apply(float& ir, float& ig, float& ib) const
+        {
+
+            float r = std::max<float>(0.0, std::min<float>(65535.0, ir));
+            float g = std::max<float>(0.0, std::min<float>(65535.0, ig));
+            float b = std::max<float>(0.0, std::min<float>(65535.0, ib));
+
+            if (r >= g) {
+                if (g > b) {
+                    RGBTone(r, g, b);     // Case 1: r >= g >  b
+                }
+                else if (b > r) {
+                    RGBTone(b, r, g);     // Case 2: b >  r >= g
+                }
+                else if (b > g) {
+                    RGBTone(r, b, g);     // Case 3: r >= b >  g
+                }
+                else {                           // Case 4: r == g == b
+                    r = lutToneCurve[r];
+                    g = lutToneCurve[g];
+                    b = g;
+                }
+            }
+            else {
+                if (r >= b) {
+                    RGBTone(g, r, b);     // Case 5: g >  r >= b
+                }
+                else if (b > g) {
+                    RGBTone(b, g, r);     // Case 6: b >  g >  r
+                }
+                else {
+                    RGBTone(g, b, r);     // Case 7: g >= b >  r
+                }
+            }
+
+            //setUnlessOOG(ir, ig, ib, r, g, b);
+            if (!(ir < 0.0 || ir > 65535.0) || !(ig < 0.0 || ig > 65535.0) || !(ib < 0.0 || ib > 65535.0)) {
```

Comment:
since this values is recurring a lot in the code maybe it would be better to have a const / const_expr declared